lib/Twig/Node/With.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+/*
+ * This file is part of Twig.
+ *
+ * (c) 2016 Fabien Potencier
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+/**
+ * Represents a nested "with" scope.
+ *
+ * @author Fabien Potencier <fabien@symfony.com>
+ */
+class Twig_Node_With extends Twig_Node
+{
+    public function __construct(Twig_Node $body, Twig_Node $variables = null, $only = false, $lineno, $tag = null)
+    {
+        $nodes = array('body' => $body);
+        if (null !== $variables) {
+            $nodes['variables'] = $variables;
+        }
+
+        parent::__construct($nodes, array('only' => (bool) $only), $lineno, $tag);
+    }
+
+    public function compile(Twig_Compiler $compiler)
+    {
+        $compiler->addDebugInfo($this);
+
+        if ($this->hasNode('variables')) {
+            $varsName = $compiler->getVarName();
+            $compiler
+                ->write(sprintf('$%s = ', $varsName))
+                ->subcompile($this->getNode('variables'))
+                ->raw(";\n")
+                ->write(sprintf("if (!is_array(\$%s)) {\n", $varsName))
+                ->indent()
+                ->write("throw new Twig_Error_Runtime('Variables passed to the \"with\" tag must be a hash.');\n")
+                ->outdent()
+                ->write("}\n")
+            ;
+
+            if ($this->getAttribute('only')) {
+                $compiler->write("\$context = array('_parent' => \$context);\n");
```

Comment:
this still gives access to the parent scope here. Shouldn't we store the old scope in a variable in the compiled template instead of storing it inside the new context ? Doing ``{% set _parent = ... %}`` inside the ``{% with %}`` allows to replace the whole scope outside the tag